cmd/encryption.go

```diff
@@ -0,0 +1,199 @@
+	// or by the server on server-side-encryption.
+	ClientSideEncryptionIV = "X-Amz-Meta-X-Amz-Iv"
+
+	// ContentEncryptionAlgorithm is the algorithm used to encrypted the content.
+	// For minio SSE-C this field will be "DARE"
+	ContentEncryptionAlgorithm = "X-Amz-Meta-X-Amz-Cek-Alg"
+)
+
+// DareHmacSha256 specifies the ContentEncryptionAlgorithm used by the server for SSE-C requests
+// and the KDF function used to generate encryption keys.
+const DareHmacSha256 = "DARE-HMAC-SHA256"
+
+// SSECustomerKeySize is the size of valid client provided encryption keys in bytes.
+// Currently AWS supports only AES256. So the SSE-C key size is fixed to 32 bytes.
+const SSECustomerKeySize = 32
+
+// SSECustomerAlgorithmAES256 the the only valid S3 SSE-C encryption algorithm.
+const SSECustomerAlgorithmAES256 = "AES256"
+
+// IsSSECustomerRequest returns true if the given HTTP header
+// contains server-side-encryption with customer provided key fields.
+func IsSSECustomerRequest(header http.Header) bool {
+	return header.Get(SSECustomerAlgorithm) != "" || header.Get(SSECustomerKey) != "" || header.Get(SSECustomerKeyMD5) != ""
+}
+
+// ParseSSECustomerRequest parses the SSE-C header fields of the provided request.
+// It returns the client provided key on success.
+func ParseSSECustomerRequest(r *http.Request) (key []byte, err error) {
+	if r.TLS == nil {
+		return nil, errInsecureSSERequest
+	}
+	header := r.Header
+	if algorithm := header.Get(SSECustomerAlgorithm); algorithm != SSECustomerAlgorithmAES256 {
+		return nil, errInvalidSSEAlgorithm
+	}
+	if header.Get(SSECustomerKey) == "" {
+		return nil, errMissingSSEKey
+	}
+	if header.Get(SSECustomerKeyMD5) == "" {
+		return nil, errMissingSSEKeyMD5
+	}
+
+	key, err = base64.StdEncoding.DecodeString(header.Get(SSECustomerKey))
+	if err != nil {
+		return nil, errInvalidSSEKey
+	}
+	if len(key) != SSECustomerKeySize {
+		return nil, errInvalidSSEKey
+	}
+
+	keyMD5, err := base64.StdEncoding.DecodeString(header.Get(SSECustomerKeyMD5))
+	if err != nil {
+		return nil, errSSEKeyMD5Mismatch
+	}
+	if md5Sum := md5.Sum(key); !bytes.Equal(md5Sum[:], keyMD5) {
+		return nil, errSSEKeyMD5Mismatch
+	}
+	return key, nil
+}
+
+// EncryptRequest takes the client provided content and encrypts the data
+// with the client provided key. It also marks the object as client-side-encrypted
+// and sets the correct headers.
+func EncryptRequest(content io.Reader, size int64, r *http.Request, metadata map[string]string, sha256sum string) (io.Reader, error) {
+	key, err := ParseSSECustomerRequest(r)
+	if err != nil {
+		return nil, err
+	}
+	nonce := make([]byte, 32) // generate random nonce to derive encryption key
+	if _, err = io.ReadFull(rand.Reader, nonce); err != nil {
+		return nil, err
+	}
+
+	mac := hmac.New(sha256.New, key)
+	mac.Write(nonce)
+	content, err = sio.EncryptReader(content, sio.Config{Key: mac.Sum(nil)})
+	if err != nil {
+		return nil, errInvalidSSEKey
```

Comment:
Is this the appropriate error response (invalid SSE Key)?

Reply:
Yes, and no - basically this should never happen because the size of the MAC is 32 bytes (SHA-256)
Currently `EncryptReader` will only return an error if the key is not 32 bytes long (all other config options are the default) which can never happen because SHA-256 produces 32 byte hash values. From my understanding the correct approach would be a panic - indicate a logical bug. However we agreed on: no panic in minio server code so it seems to be the correct error (the key has invalid size)
But we may can trace the error?